### PR TITLE
Generic filtering

### PR DESCRIPTION
**Pull request by ryandillinfelton**
This PR adds generic filtering support to xjoin-search, automating the process of adding filters for new fields in the system profile. 

When these changes are in place, any change to the system profile schema in the inventory-schemas will cause a pull request to be opened against `xjoin-search`. The automatically created PR will contain all of the necessary changes to enable filtering on allowed fields on the system profile. From there it is up to maintainers to accept or reject the changes as to not bypass existing release pipelines and procedures. I've made an effort to make the automated PRs easy to review and merge quickly, while also being safe and audit-able in the git history.

I've summarized the process below:
1. Change is merged to inventory-schemas
1. The xjoin-search workflow runs
    1. System profile schema and associated SHA are updated
    1. The update_from_schema script is run
        1. The elasticsearch mapping is updated
        1. The grapqhQL schema is updated, adding new inputs and types generated from the SPF schema
        1. The test hosts that are use in manual & automated tests are updated with mock data for new fields
        1. The test data file that holds a list of queries used in automated tests for system profile filters is updated to add test queries for the new fields
    1. A PR is opened on xjoin-search
1. The PR is reviewed and merged, the new filters follow the normal xjoin-search release pipeline.

There are a few main changes in this pull request, which I have summarized below.

1. The system profile schema, and the SHA of the commit to the [inventory-schemas](https://github.com/RedHatInsights/inventory-schemas) that schema is even with have been added to `xjoin-search` in the new `inventory-schemas` directory in `system_profile_schema.yaml` and `system_profile_schema_sha.txt` respectively.
1. The elasticsearch mapping in `seed.js` has been externalized to `test/mapping.json`.
1. A new workflow has been added `.github/workflows/update_from_schema.yml`. The workflow is triggered when an change is pushed to the master branch of the inventory-schemas repository. The workflow updates the system profile schema in xjoin-search, runs the `updateFromSchema` script, and opens a pull request with whatever changes have been made for approval by maintainers.
1. A new script `scripts/update_from_schema.ts` has been added in the new `/scripts` directory. The script parses the system profile schema and updates the following files to support new/removed fields.
    - `hosts.json` - The test hosts seed.js uses
    - `schema.graphql` - The graphQL schema
    - `test_data.json` - The queries used in the test for the auto-generated filters
    - `mapping.json` - The elasticsearch mapping used by seed.js
1. Resolvers for host resources are now generated from the system profile schema at run time. The main changes are in the following files.
   - `util/systemProfile.ts` - Functions for parsing the system_profile schema
   - `resolvers/hosts/index.ts` - Where the resolvers are generated. The previously statically defined list of partial resolver functions has been factored into two parts.
       1. The statically defined filters for data outside of the system profile facts. These are returned by `getPredefinedResolvers()` now.
       2. The generated resolvers for fields in the system profile. These are generated by and returned from `resolverMapFromSchema()`
       
A few notes:

Snapshots:
    There are some files used for testing that are snapshots of the query response. I tried to remove snapshot based tests where I could if they were effected by the changes, that way it's clear what is being tested. Some still need to use snapshots though, for those verify the response looks reasonable. I presume much of the snapshot changes are lines removed because of those test changes.

DRY: 
    There are some common functions between the `update_from_schema.ts` script, the `xjoin-search` application, particularly `system_profile.ts`. I've made an effort to insure they are in sync, but importing one from the other appears to require a fairly substantial refactor (breaking the code into 3 TS projects, scripts, app, and lib). I felt that would be a little out of scope for this PR and make this already unwieldy PR even bigger so I've chosen to forgo it at the expense of a little code duplication. If you know a better and lighter solution than the one I found let me know, otherwise I think it will be good for the time being and we can plan a follow up to address it.

System Profile Schema:
    There are some changes to how the system profile schema should look. The biggest change is the inclusion of 3 examples in the example field to be used as test data. I've talked with Asa about switching it to use the valids we use in testing in the inventory-schemas repository, however that approach presents issues with enum fields with only one possible value. The automated tests rely on being able to query just one host, out of the three so this presents some issues. Never the less I'm interested in hearing thoughts on how to improve this, it does feel a bit hacky, even with the additions to the PR checklist and README on inventory schemas. I would prefer to keep that conversation on the [inventory-schemas PR](https://github.com/RedHatInsights/inventory-schemas/pull/78) to not clog  up conversation here.
    
Comments as place markers in graphQL schema:
    There was some conversation around this in the walk-through call, but I don't believe we found a definite solution. I think the most promising option we weighed was to do something similar to the resolvers. Break out the non-SPF content of the `HostFilter` into a seperate file, or a static part of the script, then combine it with the new generate filters and replace the full contents of `HostFilter` using its start and end markers instead of comments.
    
 Tests:
     I know I was hesitant about automatically generated tests and I presume others will be too. The existing test are pretty boilerplate though so I think they are a solid 'better than nothing' the test should at least insure that there are no bugs on the happy path of a given filter. They always expect to get back the second host on the test account, which is the second host defined in `hosts.json`. The generated queries target that host, and expect to get it back. Bad test data can cause issues, but for equality as long as each test value is different it's fine. For range fields the generation predictable so it should always work. Let me know if there are specific questions about the testing. I know the generated tests have blind spots, but more specific ones can always be hand written in a follow up.
     
God speed and good luck o7.
Slack me if you have questions.

P.S. I'm trying not to balloon this any more. If there are nice to haves or improvements that can be made please consider if they would be better suited as follow ups.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,9 +12,10 @@
     "lint": "eslint 'src/**/*.ts'",
     "seed": "ts-node test/seed.js",
     "start": "tsc && LOG_PRETTY=true ts-node-dev --respawn src/app.ts",
-    "test": "npm run compile && npm run seed && NODE_ENV=test LOG_LEVEL=fatal jest -i",
+    "test": "npm run compile && npm run seed && NODE_ENV=test LOG_LEVEL=fatal jest",
```
Comment (thearifismail) [inline package.json:15]: wouldn't it be more helpful to use `debug` as the default log level for `test`

File: scripts/update_from_schema.js [hunk 2]
```diff
@@ -0,0 +1,372 @@
+// When running this script add the path to the schema file to use as argv
```
Comment (thearifismail) [inline scripts/update_from_schema.js:1]: Need a couple of sentences about why is this script needed?  The `main` function shows it updates all the major pieces.  Why is it needed now?  When to use this script?

File: scripts/update_from_schema.js [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    _.forEach(hosts, (host) => {
+        if (host["account"] == "test") {
+            test_host_ids.push(host["id"]);
+        }
+    })
+
+    return test_host_ids;
+}
+
+function updateHostTests(schema, new_host_system_profile_facts) {
```
Comment (thearifismail) [inline scripts/update_from_schema.js:298]: VSCode shows `updateHostTests` and `new_host_system_profile_facts` are defined but not used, is they really not used or get used behind the scenes.  i.e. not providing them shows errors at execution time.

File: test/index.ts [hunk 4]
```diff
@@ -6,6 +6,7 @@ import client from '../src/es';
 
 let app: any = null;
 
+jest.setTimeout(2000000)
```
Comment (thearifismail) [inline test/index.ts:9]: comment needed for time unit.

File: scripts/update_from_schema.js (hunk 3 above)
Comment (ryandillinfelton) [reply scripts/update_from_schema.js:298]: Draft PR. I've fixed this in the current local version. Some leftovers from a different approach to dynamic test creation

**Comment by fijshion** [comment]
`npm run verify` is showing a lot of lint errors.

```
✖ 220 problems (203 errors, 17 warnings)
  175 errors and 0 warnings potentially fixable with the `--fix` option.
```

**Comment by ryandillinfelton** [comment]
> `npm run verify` is showing a lot of lint errors.
> 
> ```
> ✖ 220 problems (203 errors, 17 warnings)
>   175 errors and 0 warnings potentially fixable with the `--fix` option.
> ```

All errors fixed now. There are a few warnings I'm still working out though. I think the only things I should have to change are a few of the parameter types.

**Comment by ryandillinfelton** [comment]
Okay, now `npm run verify` shouldn't have any errors or warnings. I did disable warning explicitly in a few spots  where I think it makes sense to make an exception. One was related to a generic object injection security risk. I left a comment on it explaining my reasoning on why it was incorrectly flagged, but I would appreciate a second look to make sure my assessment is correct.

Thanks for pointing out the warnings @fijshion. You should be good to review now :+1:

**Comment by fijshion** [comment]
looks good - going to wait to merge until the performance testing is complete.

**Comment by fstavela** [comment]
@fijshion How is the performance testing looking? Can we merge?

**Comment by fijshion** [comment]
@fstavela the performance tests aren't blocking. Waiting on https://github.com/RedHatInsights/xjoin-operator/pull/4 to be rebased and merged.

File: .github/workflows/update_from_schema.yaml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+          echo "::set-output name=schema_sha::$(git ls-remote https://github.com/RedHatInsights/inventory-schemas.git master | cut -c -7)"
+      - name: update schema file
+        run: |
+          cp ../inventory-schemas/schemas/system_profile/v1.yaml inventory-schemas/system_profile_schema.yaml
+          echo "${{ steps.vars.outputs.schema_sha }}" >> inventory-schemas/system_profile_schema_sha.txt
+      - name: update mapping from schema
+        run: |
+          cd scripts
+          npm ci
+          npm run updateFromSchema
```
Comment (ryandillinfelton) [inline .github/workflows/update_from_schema.yaml:38]: This line needs to be changed since the script is in typescript now

Comment (ryandillinfelton) [reply .github/workflows/update_from_schema.yaml:38]: NVM

**Review by fijshion — APPROVED** [review]

**Pull request merged**